src/sd-escape.c

```diff
@@ -0,0 +1,356 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "sd-escape.h"
+#include <errno.h>
+
+/* The following is copied from these files in systemd with minor
```

Comment:
If I add unit tests for the unit name parsing that uses these helper functions, I don't really see the point in also testing the helper functions, especially given they are directly copied from somewhere they are being tested?

Reply:
> If I add unit tests for the unit name parsing that uses these helper functions,

I don't see new unit tests in the PR, I looked 3 times, including just now. Is github's review tool hiding them from me?

> I don't really see the point in also testing the helper functions, especially given they are directly copied from somewhere they are being tested?

Because they can, and will, eventually diverge. Another option would be to export those functions from systemd so you don't have to test them in xdg-desktop-portal at all.

Reply:
> > If I add unit tests for the unit name parsing that uses these helper functions,
> 
> I don't see new unit tests in the PR, I looked 3 times, including just now. Is github's review tool hiding them from me?
> 

They don't exist yet. I was saying I can add them.

> > I don't really see the point in also testing the helper functions, especially given they are directly copied from somewhere they are being tested?
> 
> Because they can, and will, eventually diverge. Another option would be to export those functions from systemd so you don't have to test them in xdg-desktop-portal at all.

Do you mean add new library API to libsystemd upstream? That might be a good idea but I don't have time for it, sorry.

I'm not sure why they would diverge though? The copy here will become out of date when any changes are made to the copy in systemd, but I wouldn't expect us to need to make our own downstream changes here. And even if we did, the functionality we need would be tested by the higher level unit tests I am going to add.

Reply:
> Do you mean add new library API to libsystemd upstream? That might be a good idea but I don't have time for it, sorry.

https://github.com/systemd/systemd/issues/22718

> I'm not sure why they would diverge though? The copy here will become out of date when any changes are made to the copy in systemd, but I wouldn't expect us to need to make our own downstream changes here. And even if we did, the functionality we need would be tested by the higher level unit tests I am going to add.

FWIW, it always does eventually diverge. Because someone with good intentions fixes a bug they find, or simplifies the code and they don't know that it's copy/paste, and the reviewer forgets.